### PR TITLE
Modify formatting of perference's name to align vscode

### PR DESCRIPTION
**Pull request by shuyaqian**
Signed-off-by: shuyaqian <717749594@qq.com>

#### What it does

Avoid the problem that if the perference's name contains Chinese characters or special characters like C++ will be filtered.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by colin-grant-work** [comment]
@shuyaqian, thank you for your contribution! With your changes, I'm seeing some problems where names that should be split are not being split: 

image

You mention in your PR description that the problem occurs with Chinese characters and special characters like the `+` in C++. Could you add some tests to the test suite for the various cases?

For example `completePropertyWithSemicolon` should become `Complete Property With Semicolon` but `C++` should remain `C++` and `某個設定/某个设定` should remain `某個設定/某个设定`.

**Comment by shuyaqian** [comment]
> @shuyaqian, thank you for your contribution! With your changes, I'm seeing some problems where names that should be split are not being split:
> 
> image
> 
> You mention in your PR description that the problem occurs with Chinese characters and special characters like the `+` in C++. Could you add some tests to the test suite for the various cases?
> 
> For example `completePropertyWithSemicolon` should become `Complete Property With Semicolon` but `C++` should remain `C++` and `某個設定/某个设定` should remain `某個設定/某个设定`.

Thanks, I've revised it.

**Comment by colin-grant-work** [comment]
> Thanks, I've revised it.

Thank you very much! One thing I notice is that the new regex uses ranges like `[a-z]` to a lot. That means it now works well for the vanilla Latin alphabet, but it wouldn’t work for any other alphabets. To generalize it, I think it may be better to apply an iterative approach, rather than a regex, something like this (pseudocode):

```
Initialize output to empty string
Iterate over input string
If the first character can be capitalized, capitalize it.
If you encounter a capitalized character following a non-capitalized character, add a space before the capitalized character.
Otherwise just add the character

e.g.
Danish:
indstillingPåEnØ -> Indstilling På En Ø

Greek:
κάποιαΡύθμιση -> Κάποια Ρύθμιση

Russian:
некоторыеНастройки -> Некоторые Настройки

Armenian:
ինչ-որՊարամետր -> Ինչ-որ Պարամետր
```

I haven’t tested every case, but JavaScript’s `toUpperCase()` and `toLowerCase()` methods work on a wide variety of character sets, so a check like `a.toLowerCase() !== a` works pretty well to tell if something *is* capitalized, and `a.toUpperCase() !== a` works  pretty well to check whether something *could be* capitalized.

Punctuation is a challenge. For example something like `hyphenated-wordC++Setting` should maybe become `Hyphenated-work C++ Setting`, but I don’t think we have a mechanism at the moment for making any useful distinctions there. I think we should probably allow something like a `displayName` in the preference schema and only fall back to calculating it if that doesn’t exist, but that’s outside the scope of this PR.

**Comment by shuyaqian** [comment]
> > Thanks, I've revised it.
> 
> Thank you very much! One thing I notice is that the new regex uses ranges like `[a-z]` to a lot. That means it now works well for the vanilla Latin alphabet, but it wouldn’t work for any other alphabets. To generalize it, I think it may be better to apply an iterative approach, rather than a regex, something like this (pseudocode):
> 
> ```
> Initialize output to empty string
> Iterate over input string
> If the first character can be capitalized, capitalize it.
> If you encounter a capitalized character following a non-capitalized character, add a space before the capitalized character.
> Otherwise just add the character
> 
> e.g.
> Danish:
> indstillingPåEnØ -> Indstilling På En Ø
> 
> Greek:
> κάποιαΡύθμιση -> Κάποια Ρύθμιση
> 
> Russian:
> некоторыеНастройки -> Некоторые Настройки
> 
> Armenian:
> ինչ-որՊարամետր -> Ինչ-որ Պարամետր
> ```
> 
> I haven’t tested every case, but JavaScript’s `toUpperCase()` and `toLowerCase()` methods work on a wide variety of character sets, so a check like `a.toLowerCase() !== a` works pretty well to tell if something _is_ capitalized, and `a.toUpperCase() !== a` works pretty well to check whether something _could be_ capitalized. I’ve got some code [here](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBBCcQQ0SLRABCkAeBA2AGEGIJIK4ALygAAUDHy11QlE4YwAPqAuNxIMkMlyAJRCkHwRpebkAPlAAG9FCUSk4CWROqAAITc3nS2Xy1wAMltYgC6GZbI5kD5CrNFrEAG5FABfSnwanpQQun7szk8-mC4WiiVSnjWziK5XwVUedVa3X60CGwmdc3cq1ylN2h0MJ3wF7wKRkSPuz1F30B4LuGgpiTEEWjUC8gDk-b9VJpSWi3fjfdA-cA1qCAN1BAJ6gpUAHqCAANBAFKggENQVebwAuoIBTUEA-qCroeKdtuDmoF6QVAKMh+xTJeA4vnnihTgAMPpEAB5VdwJ1GAJ3DFWJvwoABqCDPRzfVEX5d8LV5D9QHtdEom4Rlw3rN1uD5f9ALFABtCgAF0FRgvVc31TtCLGCDeQIntiLIqtsIbD0-Vzf0IWgTkRFSPkhHY3D8PHZjmBI8jUIdU1hJZCNRMYGlr1vH4KJ1KjqMGSAuwk0AGOnIyILHACJKkrj9R4mB+Ng7TaP0wymPjCytNAQNc2Uq8bzvKdnNGVzA0UEcM2A+BmDEsz434Bzop0vTJyQkp+0AblBFxXVdQC3HcD2PfsFQfFQcAaDxjVARBn2cEF2Bob4LzFdgGGYSB+FgaB1ESBgQT0EQvAAIgAUVWBhEBqfQ+qSHE+vABxUH7YQADldL6oIVAAeW+Mhdk5fhmuOVBHAUVxpDwPNIAqnRe2+fiEEQGFhCuOqJGfaIRAMDgcW6Fg2GlIISiAA) that works on the Armenian example, and should cover the rest, though it can certainly be improved.
> 
> Punctuation is a challenge. For example something like `hyphenated-wordC++Setting` should maybe become `Hyphenated-word C++ Setting`, but I don’t think we have a mechanism at the moment for making any useful distinctions there. I think we should probably allow something like a `displayName` in the preference schema and only fall back to calculating it if that doesn’t exist, but that’s outside the scope of this PR.


> > Thanks, I've revised it.
> 
> Thank you very much! One thing I notice is that the new regex uses ranges like `[a-z]` to a lot. That means it now works well for the vanilla Latin alphabet, but it wouldn’t work for any other alphabets. To generalize it, I think it may be better to apply an iterative approach, rather than a regex, something like this (pseudocode):
> 
> ```
> Initialize output to empty string
> Iterate over input string
> If the first character can be capitalized, capitalize it.
> If you encounter a capitalized character following a non-capitalized character, add a space before the capitalized character.
> Otherwise just add the character
> 
> e.g.
> Danish:
> indstillingPåEnØ -> Indstilling På En Ø
> 
> Greek:
> κάποιαΡύθμιση -> Κάποια Ρύθμιση
> 
> Russian:
> некоторыеНастройки -> Некоторые Настройки
> 
> Armenian:
> ինչ-որՊարամետր -> Ինչ-որ Պարամետր
> ```
> 
> I haven’t tested every case, but JavaScript’s `toUpperCase()` and `toLowerCase()` methods work on a wide variety of character sets, so a check like `a.toLowerCase() !== a` works pretty well to tell if something _is_ capitalized, and `a.toUpperCase() !== a` works pretty well to check whether something _could be_ capitalized. I’ve got some code [here](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBBCcQQ0SLRABCkAeBA2AGEGIJIK4ALygAAUDHy11QlE4YwAPqAuNxIMkMlyAJRCkHwRpebkAPlAAG9FCUSk4CWROqAAITc3nS2Xy1wAMltYgC6GZbI5kD5CrNFrEAG5FABfSnwanpQQun7szk8-mC4WiiVSnjWziK5XwVUedVa3X60CGwmdc3cq1ylN2h0MJ3wF7wKRkSPuz1F30B4LuGgpiTEEWjUC8gDk-b9VJpSWi3fjfdA-cA1qCAN1BAJ6gpUAHqCAANBAFKggENQVebwAuoIBTUEA-qCroeKdtuDmoF6QVAKMh+xTJeA4vnnihTgAMPpEAB5VdwJ1GAJ3DFWJvwoABqCDPRzfVEX5d8LV5D9QHtdEom4Rlw3rN1uD5f9ALFABtCgAF0FRgvVc31TtCLGCDeQIntiLIqtsIbD0-Vzf0IWgTkRFSPkhHY3D8PHZjmBI8jUIdU1hJZCNRMYGlr1vH4KJ1KjqMGSAuwk0AGOnIyILHACJKkrj9R4mB+Ng7TaP0wymPjCytNAQNc2Uq8bzvKdnNGVzA0UEcM2A+BmDEsz434Bzop0vTJyQkp+0AblBFxXVdQC3HcD2PfsFQfFQcAaDxjVARBn2cEF2Bob4LzFdgGGYSB+FgaB1ESBgQT0EQvAAIgAUVWBhEBqfQ+qSHE+vABxUH7YQADldL6oIVAAeW+Mhdk5fhmuOVBHAUVxpDwPNIAqnRe2+fiEEQGFhCuOqJGfaIRAMDgcW6Fg2GlIISiAA) that works on the Armenian example, and should cover the rest, though it can certainly be improved.
> 
> Punctuation is a challenge. For example something like `hyphenated-wordC++Setting` should maybe become `Hyphenated-word C++ Setting`, but I don’t think we have a mechanism at the moment for making any useful distinctions there. I think we should probably allow something like a `displayName` in the preference schema and only fall back to calculating it if that doesn’t exist, but that’s outside the scope of this PR.

I have a preference for 'SecSolar', which is an acronym, and I don't want it to be 'Sec Solar', but in VSCODE it's also 'Sec Solar'. I think this can be aligned with vscode and ignored for the time being. The rest of the logic is executed according to the pseudocode above.

**Comment by shuyaqian** [comment]
> > Thanks, I've revised it.
> 
> Thank you very much! One thing I notice is that the new regex uses ranges like `[a-z]` to a lot. That means it now works well for the vanilla Latin alphabet, but it wouldn’t work for any other alphabets. To generalize it, I think it may be better to apply an iterative approach, rather than a regex, something like this (pseudocode):
> 
> ```
> Initialize output to empty string
> Iterate over input string
> If the first character can be capitalized, capitalize it.
> If you encounter a capitalized character following a non-capitalized character, add a space before the capitalized character.
> Otherwise just add the character
> 
> e.g.
> Danish:
> indstillingPåEnØ -> Indstilling På En Ø
> 
> Greek:
> κάποιαΡύθμιση -> Κάποια Ρύθμιση
> 
> Russian:
> некоторыеНастройки -> Некоторые Настройки
> 
> Armenian:
> ինչ-որՊարամետր -> Ինչ-որ Պարամետր
> ```
> 
> I haven’t tested every case, but JavaScript’s `toUpperCase()` and `toLowerCase()` methods work on a wide variety of character sets, so a check like `a.toLowerCase() !== a` works pretty well to tell if something _is_ capitalized, and `a.toUpperCase() !== a` works pretty well to check whether something _could be_ capitalized. I’ve got some code [here](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBBCcQQ0SLRABCkAeBA2AGEGIJIK4ALygAAUDHy11QlE4YwAPqAuNxIMkMlyAJRCkHwRpebkAPlAAG9FCUSk4CWROqAAITc3nS2Xy1wAMltYgC6GZbI5kD5CrNFrEAG5FABfSnwanpQQun7szk8-mC4WiiVSnjWziK5XwVUedVa3X60CGwmdc3cq1ylN2h0MJ3wF7wKRkSPuz1F30B4LuGgpiTEEWjUC8gDk-b9VJpSWi3fjfdA-cA1qCAN1BAJ6gpUAHqCAANBAFKggENQVebwAuoIBTUEA-qCroeKdtuDmoF6QVAKMh+xTJeA4vnnihTgAMPpEAB5VdwJ1GAJ3DFWJvwoABqCDPRzfVEX5d8LV5D9QHtdEom4Rlw3rN1uD5f9ALFABtCgAF0FRgvVc31TtCLGCDeQIntiLIqtsIbD0-Vzf0IWgTkRFSPkhHY3D8PHZjmBI8jUIdU1hJZCNRMYGlr1vH4KJ1KjqMGSAuwk0AGOnIyILHACJKkrj9R4mB+Ng7TaP0wymPjCytNAQNc2Uq8bzvKdnNGVzA0UEcM2A+BmDEsz434Bzop0vTJyQkp+0AblBFxXVdQC3HcD2PfsFQfFQcAaDxjVARBn2cEF2Bob4LzFdgGGYSB+FgaB1ESBgQT0EQvAAIgAUVWBhEBqfQ+qSHE+vABxUH7YQADldL6oIVAAeW+Mhdk5fhmuOVBHAUVxpDwPNIAqnRe2+fiEEQGFhCuOqJGfaIRAMDgcW6Fg2GlIISiAA) that works on the Armenian example, and should cover the rest, though it can certainly be improved.
> 
> Punctuation is a challenge. For example something like `hyphenated-wordC++Setting` should maybe become `Hyphenated-word C++ Setting`, but I don’t think we have a mechanism at the moment for making any useful distinctions there. I think we should probably allow something like a `displayName` in the preference schema and only fall back to calculating it if that doesn’t exist, but that’s outside the scope of this PR.

I have a preference for 'SecSolar', which is an acronym, and I don't want it to be 'Sec Solar', but in VSCODE it's also 'Sec Solar'. I think this can be aligned with vscode and ignored for the time being. The rest of the logic is executed according to the pseudocode above.

File: packages/preferences/src/browser/util/preference-tree-generator.ts
Comment (colin-grant-work) [inline packages/preferences/src/browser/util/preference-tree-generator.ts:139]:
```suggestion
    protected formatString(string: string): string {
```

File: packages/preferences/src/browser/util/preference-tree-generator.ts
Comment (colin-grant-work) [inline packages/preferences/src/browser/util/preference-tree-generator.ts:150]:
```suggestion
    protected isUpperCase(char: string): boolean {
```

**Review by colin-grant-work — APPROVED** [review]
The code looks good and the tests are a very nice addition - thank you!

**Review by colin-grant-work — CHANGES_REQUESTED** [review]
Oops, a test is now failing. Looks like it expects capitalization after white space, as well as string-initially. I think that expectation is incorrect. If you create a preference name in the following form:

```
"references.My perfectly formatted Display Name"`
```
VSCode does not capitalize the words after spaces:

image

So you can simply change the expected output.

**Comment by shuyaqian** [comment]
Thanks, I've revised it.  I don't know why the test cases don't run automatically after force-push🤪，maybe need you to approve.

**Comment by colin-grant-work** [comment]
@shuyaqian, looks like it did re-run, and there's still a tiny problem: the current code is adding a space in the sequence `space + capital`:

```
@theia/preferences:   1) preference-tree-generator
@theia/preferences:        PreferenceTreeGenerator.format:
@theia/preferences:       AssertionError: expected 'Aaa Bbb Ccc  Dddd eee' to equal 'Aaa Bbb Ccc Dddd eee'
@theia/preferences:       + expected - actual
@theia/preferences:       -Aaa Bbb Ccc  Dddd eee      (colin-grant-work: two spaces in 'Ccc  Dddd' here instead of the expected one)
@theia/preferences:       +Aaa Bbb Ccc Dddd eee
```
You can also run the tests locally: from the `packages/preferences` directory, run `yarn test`, and it will run the tests for that package.

**Comment by shuyaqian** [comment]
> Thanks, I've revised it. I don't know why the test cases don't run automatically after force-push🤪，maybe need you to approve.

I ignored the space in front of the uppercase letter, and I modified the criteria, now it's okay.

**Comment by shuyaqian** [comment]
> @shuyaqian, looks like it did re-run, and there's still a tiny problem: the current code is adding a space in the sequence `space + capital`:
> 
> ```
> @theia/preferences:   1) preference-tree-generator
> @theia/preferences:        PreferenceTreeGenerator.format:
> @theia/preferences:       AssertionError: expected 'Aaa Bbb Ccc  Dddd eee' to equal 'Aaa Bbb Ccc Dddd eee'
> @theia/preferences:       + expected - actual
> @theia/preferences:       -Aaa Bbb Ccc  Dddd eee      (colin-grant-work: two spaces in 'Ccc  Dddd' here instead of the expected one)
> @theia/preferences:       +Aaa Bbb Ccc Dddd eee
> ```
> 
> You can also run the tests locally: from the `packages/preferences` directory, run `yarn test`, and it will run the tests for that package.

I ignored the space in front of the uppercase letter, and I modified the criteria, now it's okay.

File: packages/preferences/src/browser/util/preference-tree-generator.ts
Comment (colin-grant-work) [inline packages/preferences/src/browser/util/preference-tree-generator.ts:142]: Minor: I think this is fine as is - I'll approve it momentarily - but the check `!== ' '` could be generalized to regex `/\s/`, since not all scripts use the ' ' character.

**Review by colin-grant-work — APPROVED** [review]
Tests are passing and the code looks good. Thanks for the contribution and for the tests, @shuyaqian!

**Comment by colin-grant-work** [comment]
I will merge this tomorrow, if there are no objections.

**Comment by shuyaqian** [comment]
> I will merge this tomorrow, if there are no objections.

Thanks, hopefully this will end the PR this time😅

**Comment by colin-grant-work** [comment]
> Thanks, hopefully this will end the PR this time😅

Thanks for going along with my expansion of the scope of the PR. I think this is a big improvement over the previous implementation.

**Pull request merged**
